mlir/lib/Conversion/TosaToRock/TosaToRock.cpp

```diff
@@ -426,18 +426,70 @@ struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
   }
 };
 
+template <typename TosaReduceOp>
+typename std::enable_if_t<
+    std::is_same<TosaReduceOp, tosa::ReduceSumOp>::value ||
+        std::is_same<TosaReduceOp, tosa::ReduceMaxOp>::value,
+    LogicalResult> static matchAndRewriteReductions(TosaReduceOp op,
+                                                    rock::ReduceMethod rMethod,
+                                                    Attribute outputInitVal,
+                                                    ConversionPatternRewriter
+                                                        &rw) {
+  Location loc = op->getLoc();
+  auto outputType = op.getType().template cast<RankedTensorType>();
+  Value output =
+      rw.create<bufferization::AllocTensorOp>(loc, outputType, ValueRange{});
+  StringAttr arch;
+  Optional<uint32_t> num_cu;
+  rock::GemmFeatures features;
+  std::tie(arch, num_cu, features) = getArchAttributes(op);
+
+  int32_t blockSize = 256;
+  auto elementCount =
+      op.getInput().getType().template cast<ShapedType>().getNumElements();
+  int32_t gridSize = (elementCount + blockSize - 1) / blockSize;
+  if (num_cu.has_value()) {
+    gridSize = std::max((int32_t)(20 * num_cu.value()), gridSize);
```

Comment:
Shouldn't this, perhaps, be a min()?
Otherwise we're launching redundant work?

Reply:
you are right! fixed now.